### PR TITLE
`Kernel.linop` method

### PR DESCRIPTION
**Pull request by marvinpfoertner**
# In a Nutshell
This PR introduces a `linop` method in `Kernel`, which allows for efficient representations of structured kernel matrices.

# Detailed Description
- `Kernel.linop`
- `Kernel.matrix` now behaves like `Kernel.linop(...).todense()`
- the `output_shape` of a `Kernel` is now split into `output_shape_0` and `output_shape_1`

# Related Issues
None

### CODE REVIEW

File: src/probnum/randprocs/_random_process.py [hunk 1]
```diff
@@ -105,11 +105,15 @@ def __init__(
...
+                cov.output_shape_0 == self._output_shape
+                and cov.output_shape_1 == self._output_shape
+            ):
                 raise ValueError(
-                    f"The `output_shape` of the covariance function must be given by "
-                    f"`2 * self.output_shape` (`{cov.output_shape}` != "
-                    f"`{2 * self._output_shape}`)."
+                    f"Both `output_shape_0` and `output_shape_1` of the covariance "
+                    f"function must be set to the `output_shape` {self.output_shape} "
+                    f"of the random process, but got kernel with output shapes "
```
Comment (JonathanWenger) [inline src/probnum/randprocs/_random_process.py:115]:
```suggestion
                    f"of the random process, but got covariance function with output shapes "
```

File: src/probnum/randprocs/kernels/_kernel.py [hunk 2]
```diff
@@ -273,82 +278,136 @@ def matrix(
...
-            function(s) on ``x0`` and ``x1``.
-            Depending on the shape of the inputs, ``batch_shape`` is either ``(M, N)``,
-            ``(M,)``, ``(N,)``, or ``()``.
+        k_x0_x1
+            *shape=* ``(``\ :attr:`output_size_0` ``* N0,`` :attr:`output_size_1`
+            ``* N1)``
+            *with* ``N0 = prod(batch_shape_0)`` and ``N1 = prod(batch_shape_1)`` --
+            The covariance matrix corresponding to the given batches of input points.
+            The order of the rows and columns of the covariance matrix corresponds to
+            the order of entries obtained by flattening :class:`~numpy.ndarray`\\ s with
```
Comment (JonathanWenger) [inline src/probnum/randprocs/kernels/_kernel.py:304]:
```suggestion
            the order of entries obtained by flattening :class:`~numpy.ndarray`\ s with
```

File: src/probnum/randprocs/kernels/_kernel.py [hunk 3]
```diff
@@ -273,82 +278,136 @@ def matrix(
...
+        particularly when using iterative methods to solve the associated linear
+        systems.
 
-        return self(x0, x1)
+        For instance, covariance matrices induced by separable covariance functions
+        (e.g. tensor products of covariance functions or separable multi-output kernels)
+        can often be represented as :class:`~probnum.linops.KroneckerProduct`\ s of
+        smaller covariance matrices and frameworks like :mod:`pykeops<pykeops.numpy>`
+        can be used to implement efficient matrix-vector products with covariance
+        matrices without needing to construct the entire matrix in memory.
```
Comment (JonathanWenger) [inline src/probnum/randprocs/kernels/_kernel.py:349]: Love the detailed docstring here why `.linop` exists and what it is used for.

File: src/probnum/randprocs/kernels/_kernel.py [hunk 4]
```diff
@@ -273,82 +278,136 @@ def matrix(
...
+        x0: np.ndarray,
+        x1: Optional[np.ndarray],
     ) -> np.ndarray:
-        """Implementation of the kernel evaluation which is called after input checking.
-
-        When implementing a particular kernel, the subclass should implement the kernel
-        computation by overwriting this method. It is called by the :meth:`__call__`
-        method after applying input checking. The implementation must return the array
-        described in the "Returns" section of the :meth:`__call__` method.
+        """Implementation of the covariance function, evaluation which is called after
```
Comment (JonathanWenger) [inline src/probnum/randprocs/kernels/_kernel.py:403]:
```suggestion
        """Implementation of the covariance function evaluation which is called after
```

**Review by JonathanWenger — CHANGES_REQUESTED** [review]
Looks good! Just a few docstring fixes.

File: src/probnum/randprocs/kernels/_kernel.py (hunk 3 above)
Comment (marvinpfoertner) [reply src/probnum/randprocs/kernels/_kernel.py:349]: 😁 🙏🏻

**Review by JonathanWenger — APPROVED** [review]
👍👍👍

**Pull request merged**